server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamState.java

```diff
@@ -38,30 +46,45 @@ public UpstreamState(Upstream upstream, int errorsThreshold) {
      * @param status response status code from upstream
      * @param retryAfterSeconds time in seconds when upstream may become available; only take into account with 429 status code
      */
-    public synchronized void failed(HttpStatus status, long retryAfterSeconds) {
+    void fail(HttpStatus status, long retryAfterSeconds) {
+        this.source = retryAfterSeconds == -1 ? RetryAfterSource.CORE : RetryAfterSource.UPSTREAM;
+        this.status = status;
         if (status == HttpStatus.TOO_MANY_REQUESTS) {
-            retryAfter = System.currentTimeMillis() + Math.max(retryAfterSeconds, 0) * 1000;
+            retryAfterSeconds = source == RetryAfterSource.CORE ? DEFAULT_RETRY_AFTER_SECONDS_VALUE : retryAfterSeconds;
+            setReplyAfter(retryAfterSeconds);
             log.warn("Upstream {} limit hit: retry after {}", upstream.getEndpoint(), Instant.ofEpochMilli(retryAfter).toString());
-        }
-
-        if (status.is5xx()) {
-            if (++errorCount >= errorsThreshold) {
-                retryAfter = System.currentTimeMillis()
-                             + Math.min(INITIAL_BACKOFF_DELAY_MS * (1L << errorCount), MAX_BACKOFF_DELAY_MS);
+        } else if (status.is5xx()) {
+            if (source == RetryAfterSource.CORE) {
+                errorCount++;
+                if (errorCount >= errorsThreshold) {
+                    // exponential or max backoff in seconds
+                    retryAfterSeconds = Math.min(1L << errorCount, MAX_BACKOFF_DELAY_SEC);
```

Comment:
Could give a negative, e.g. (1 << 31): -2147483648

Reply:
I think we can remove `errorCount` and set retryAfter immediately as we do 429

Reply:
fixed